cartographer/mapping/map_builder.cc

```diff
@@ -0,0 +1,132 @@
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+#include "cartographer/mapping/map_builder.h"
+
+#include <cmath>
+#include <limits>
+#include <memory>
+#include <unordered_map>
+#include <utility>
+
+#include "cartographer/common/make_unique.h"
+#include "cartographer/mapping_2d/global_trajectory_builder.h"
+#include "cartographer/mapping_3d/global_trajectory_builder.h"
+#include "cartographer/mapping_3d/local_trajectory_builder_options.h"
+#include "cartographer/sensor/laser.h"
+#include "cartographer/sensor/voxel_filter.h"
+#include "cartographer/transform/rigid_transform.h"
+#include "cartographer/transform/transform.h"
+#include "glog/log_severity.h"
+#include "glog/logging.h"
+
+namespace cartographer {
+namespace mapping {
+
+proto::MapBuilderOptions CreateMapBuilderOptions(
+    common::LuaParameterDictionary* const parameter_dictionary) {
+  proto::MapBuilderOptions options;
+  options.set_use_trajectory_builder_2d(
+      parameter_dictionary->GetBool("use_trajectory_builder_2d"));
+  *options.mutable_trajectory_builder_2d_options() =
+      mapping_2d::CreateLocalTrajectoryBuilderOptions(
+          parameter_dictionary->GetDictionary("trajectory_builder_2d").get());
+  options.set_use_trajectory_builder_3d(
+      parameter_dictionary->GetBool("use_trajectory_builder_3d"));
+  *options.mutable_trajectory_builder_3d_options() =
+      mapping_3d::CreateLocalTrajectoryBuilderOptions(
+          parameter_dictionary->GetDictionary("trajectory_builder_3d").get());
+  options.set_num_background_threads(
+      parameter_dictionary->GetNonNegativeInt("num_background_threads"));
+  *options.mutable_sparse_pose_graph_options() =
+      CreateSparsePoseGraphOptions(
+          parameter_dictionary->GetDictionary("sparse_pose_graph").get());
+  CHECK_NE(options.use_trajectory_builder_2d(), options.use_trajectory_builder_3d());
+  return options;
+}
+
+MapBuilder::MapBuilder(
+    const proto::MapBuilderOptions& options,
+    std::deque<TrajectoryNode::ConstantData>* constant_node_data)
+    : options_(options),
+      thread_pool_(options.num_background_threads()),
+      sparse_pose_graph_2d_(options.use_trajectory_builder_2d()
+                                ? new mapping_2d::SparsePoseGraph(
+                                      options_.sparse_pose_graph_options(),
+                                      &thread_pool_, constant_node_data)
+                                : nullptr),
+      sparse_pose_graph_3d_(options.use_trajectory_builder_3d()
+                                ? new mapping_3d::SparsePoseGraph(
+                                      options_.sparse_pose_graph_options(),
+                                      &thread_pool_, constant_node_data)
+                                : nullptr),
+      sparse_pose_graph_(
+          options.use_trajectory_builder_2d()
```

Comment:
Did you consider moving this and maybe the construction above out of the initializer list and assign the pointer. This would get rid of the slightly ugly static_cast: Like
if (...use...2d) { ...2d = make_unique.... 
  pose_graph = ....
}
if (...3d) ...
etc